ffi/src/handle.rs

```diff
@@ -1,235 +1,616 @@
+        ///   reference. In particular, this means the caller must ensure no other thread can access this
+        ///   handle while the returned reference is still alive.
+        pub unsafe fn as_mut(&mut self) -> &mut T {
+            H::as_mut(self.ptr.cast().as_ptr())
+        }
+    }
+
+    // [`Handle`] operations applicable only to shared handles, with implementations forwarded to
+    // the appropriate specialization of `MutableHandleOps`.
+    impl<T, S, H> Handle<H>
+    where
+        T: ?Sized,
+        H: HandleDescriptor<Target = T, Mutable = False, Sized = S>
+            + SharedHandleOps<T, S, From = Arc<T>>,
+    {
+        /// Clones this shared handle, increasing the underlying object's refcount by one. Unsafe
+        /// equivalent to [`Clone::clone`]. The new handle is independent of the source handle, and
+        /// should be dropped when it is no longer needed. This is equivalent to calling
+        /// `clone_as_arc().into_handle()`.
+        ///
+        /// # Safety
+        ///
+        /// Caller asserts that the handle is [valid][Handle#Validity].
+        pub unsafe fn clone_handle(&self) -> Self {
+            self.clone_as_arc().into()
+        }
+
+        /// Clones a new `Arc<T>` out of this handle, increasing the underlying object's refcount by
+        /// one. This is equivalent to calling `clone_handle().into_inner()`.
+        ///
+        /// # Safety
+        ///
+        /// Caller asserts that the handle is [valid][Handle#Validity].
+        pub unsafe fn clone_as_arc(&self) -> Arc<T> {
+            H::clone_arc(self.ptr.cast().as_ptr())
+        }
+    }
+
+    // NOTE: While it would be really convenient to `impl From<T> for Handle<H>`, that definition is
+    // not disjoint and conflicts with the standard `impl From<T> for T` (since there's no way to
+    // prove categorically that T != H). Ditto for `impl<X: Into<Box<T>>> From<X> for Handle<H>`.
+    impl<T, S, H> From<Box<T>> for Handle<H>
+    where
+        T: ?Sized,
+        H: HandleDescriptor<Target = T, Mutable = True, Sized = S>
+            + MutableHandleOps<T, S, From = Box<T>>,
+    {
+        fn from(val: Box<T>) -> Handle<H> {
+            let ptr = H::into_handle_ptr(val).cast();
+            Handle { ptr }
+        }
+    }
+
+    impl<T, S, H> From<Arc<T>> for Handle<H>
+    where
+        T: ?Sized,
+        H: HandleDescriptor<Target = T, Mutable = False, Sized = S>
+            + SharedHandleOps<T, S, From = Arc<T>>,
+    {
+        fn from(val: Arc<T>) -> Handle<H> {
+            let ptr = H::into_handle_ptr(val).cast();
+            Handle { ptr }
+        }
+    }
+
+    pub trait MutableHandle {}
+    impl<T, S, H> MutableHandle for H
+    where
+        T: ?Sized,
+        H: HandleDescriptor<Target = T, Mutable = True, Sized = S>,
+    {
+    }
+
+    pub trait SharedHandle {}
+    impl<T, S, H> SharedHandle for H
+    where
+        T: ?Sized + Sync,
```

Comment:
All handles are `Send`, but shared handles also need to be `Sync` because the engine is free to pass them as args to kernel functions from any number of threads at the same time.